packages/typespec-client-generator-core/src/internal-utils.ts

```diff
@@ -593,3 +595,45 @@ export function getValueTypeValue(
       return undefined;
   }
 }
+
+/**
+ * Get all of the root user-defined namespaces in a TypeSpec definition if they're tied to a client
+ * @param context
+ * @returns
+ */
+export function getRootUserDefinedNamespaceName(
+  context: TCGCContext,
+): [string, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  const rootNamespaces = [...context.program.getGlobalNamespaceType().namespaces.values()];
+
+  // We explicitly want namespaces that are used to define a client, not namespaces that define the service of a client
+  const clientNamespaces = listClients(context).map((x) =>
+    x.type.kind === "Namespace" ? x.type : x.type.namespace,
+  );
+  const globalNamespaces: string[] = [];
+  for (const namespace of clientNamespaces) {
+    const segments = [];
+    let currNamespace: Namespace | undefined = namespace;
+    while (currNamespace) {
+      segments.unshift(currNamespace.name);
+      if (rootNamespaces.includes(currNamespace)) {
+        globalNamespaces.push(segments.join("."));
+        break;
+      }
+      currNamespace = currNamespace.namespace;
+    }
+  }
+  // if we override with namespace flag, we should override the global namespace to the namespace flag
+  if (globalNamespaces.length !== 1) {
```

Comment:
I still not sure whether I understand what the logic for.

If we have 2 client, one on TypeSpec namespace `Service.Client1`, anther on `Service.Client2`, the logic here would anyway produce a `globalNamespaces` with 2 items (`rootNamespaces` would contain 1 `Service`)?